src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/RazorDynamicFileInfoProvider.cs

```diff
@@ -67,26 +68,46 @@ public RazorDynamicFileInfoProvider(Lazy<RazorWorkspaceListenerInitializer> razo
         Contract.ThrowIfNull(LanguageServerHost.Instance, "We don't have an LSP channel yet to send this request through.");
         var clientLanguageServerManager = LanguageServerHost.Instance.GetRequiredLspService<IClientLanguageServerManager>();
 
-        var response = await clientLanguageServerManager.SendRequestAsync<ProvideDynamicFileParams, ProvideDynamicFileResponse>(
+        var response = await clientLanguageServerManager.SendRequestAsync<RazorProvideDynamicFileParams, RazorProvideDynamicFileResponse>(
             ProvideRazorDynamicFileInfoMethodName, requestParams, cancellationToken);
 
-        // Since we only sent one file over, we should get either zero or one URI back
-        var responseUri = response.CSharpDocument?.Uri;
-
-        if (responseUri == null)
+        if (response.CSharpDocument is null)
         {
             return null;
         }
-        else
+
+        // Since we only sent one file over, we should get either zero or one URI back
+        var responseUri = response.CSharpDocument.Uri;
+        var dynamicFileInfoFilePath = ProtocolConversions.GetDocumentFilePathFromUri(responseUri);
+
+        if (response.Edits is not null)
         {
-            var dynamicFileInfoFilePath = ProtocolConversions.GetDocumentFilePathFromUri(responseUri);
-            return new DynamicFileInfo(dynamicFileInfoFilePath, SourceCodeKind.Regular, EmptyStringTextLoader.Instance, designTimeOnly: true, documentServiceProvider: null);
+            var project = _workspaceFactory.Workspace.CurrentSolution.GetRequiredProject(projectId);
+            var document = project.Documents.FirstOrDefault(
+                d => d.FilePath is not null && PathUtilities.PathsEqual(d.FilePath, dynamicFileInfoFilePath));
+
+            var textChanges = response.Edits.Select(e => new TextChange(e.Span.ToTextSpan(), e.NewText));
+            var textLoader = new TextChangesTextLoader(document, textChanges);
```

Comment:
how often does this happen?  Wondering if we need to instantiate a new loader each time or there was one TextLoader that we updated with the latest edits

Reply:
For closed documents it shouldn't be too often. It would be based on how often we have a diff when we generate. The most immediate cause I can think of would be if project info changed (going from unknown project to known project).